File: scala/src/main/scala/fun/pierre/lists/P01.scala
package fun.pierre.lists

import java.util.NoSuchElementException
import scala.annotation.tailrec

object P01 {

  /*
  Using the built in last function of a list
  */
  def last[T](list: List[T]): T = list.last

  /*
  Using tail recursion we ignore the first element until we have a list with only one element
  */
  @tailrec
  def lastRecursion[T](list: List[T]): T = list match {
    case x :: Nil => x
    case _ :: xs => lastRecursion(xs)
    case _ => throw new NoSuchElementException("list is empty")
  }

}

File: java/src/main/java/fun/pierre/lists/P01.java
package fun.pierre.lists;

import java.util.List;

public class P01 {
    /**
     * Return last element of a list
     * @param _list non-empty list
     * @param <T> type of element in list
     * @return last element in list
     */
    public static<T> T last(List<T> _list) {
        int size = _list.size();
        if (size == 0) {
            throw new IllegalArgumentException("Empty List");
        }
        return _list.get(size - 1);
    }
}
